lib/Models/RegionProvider.js

```diff
@@ -0,0 +1,405 @@
+        return true; // unambiguous successful match
+    }
+
+    if (defined(disambigId) && !defined(disambigCode)) {
+        // warn? we have a value to disambiguate against, but it's not provided
+        return true;
+    }
+
+    // Hooray, let's disambiguate. Our args look something like: Campbelltown, Campbelltown, New South Wales, NSW.
+
+    var dabMatched = this._getDisambigProvider().codeMatchesRegionID(disambigId, disambigCode);
+    if (!dabMatched) {
+        console.log ('"' + id + '" matched "' + code + '", but "' + disambigId + '" didn\'t match "' + disambigCode + '"');
+    }
+    return dabMatched;
+};
+
+/**
+ * Given a region code, try to find a region that matches it, using replacements, disambiguation and other wizardry.
+ * @param {String} code Code to search for.
+ * @returns {Number} Zero-based index in list of regions if successful, or -1.
+ */
+RegionProvider.prototype.findRegionIndex = function(code, disambigCode) { //, ids, rd, row) {
+    for (var i = 0; i < this.regions.length; i++) {
+        var dabId;
+        if (defined(this.disambigColumns) && defined(this.disambigColumns[0])) {
+            dabId =  this.regions[i][this.disambigColumns[0].regionProp];    
+        }
+        if (this.codeMatchesRegionID(this.regions[i].id, code, dabId, disambigCode)) {
+            return i;
+        }
+    }
+    return -1;
+};
+
+/**
+ *   Provides a list of values for each region, based on a provided dataset.
+ *
+ *   @param {Dataset} dataset
+ *   @param {String} regionVariable
+ *   @param {Array} rowList Optional array of integers: only show these rows from dataset. Used for time-filtering.
+ *   @returns {Object} Object containing properties:
+ *      regionValues: Sparse array of values corresponding to the provided list.
+ *      ambiguousMatches: Object whose keys are IDs of regions that had more than one row match against them.
+ *      failedMatches: Object whose keys are unmatched region codes.
+ *      successes: Number of successful matches.
+ *      totalRows: Number of attempted matches.
+ */
+RegionProvider.prototype.getRegionValues = function (dataset, regionVariable, disambigVariable, rowList) {
+    if (this.regions.length < 1) {
+        // should this return a promise instead of throwing?
+            throw new DeveloperError('Region provider not ready to match regions.');
+    }
+    var codes = dataset.getVariableEnums(regionVariable);
+    var disambigCodes = (defined(disambigVariable) ? dataset.getVariableEnums(disambigVariable) : []);
+    var vals = dataset.getVariableValues(dataset.getDataVariable());
+    if (!defined(vals)) {
+        console.log('Warning: no data variable for region-mapped dataset with region variable "' + regionVariable + '"');
+    }
+
+    var numericCodes = false;
+    if (!defined(codes)) {
+        // not an enum type? the IDs must be digit-based and have been misclassified as numbers
+        // ##TODO think about whether this matters for disambigs
```

Comment:
Let's try to avoid TODOs in the code.  This sounds like something that should be considered before we merge this.  In other cases, TODOs can be turned into GitHub issues instead.